torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -281,10 +292,15 @@ std::shared_ptr<FutureMessage> ProcessGroupAgent::send(
             std::chrono::steady_clock::now().time_since_epoch());
     {
       std::lock_guard<std::mutex> lock{futureMutex_};
-      futures_[requestId] = std::make_pair(future, futureStartTime);
+      futures_[requestId] = FutureInfo(future, futureStartTime, to.id_);
       // insert future into timeouts map to keep track of its timeout
       futureTimeouts_[futureStartTime].push_back(requestId);
+      // Signal watchdog for future timeouts to begin
+      if (futures_.size() == 1) {
+        futureTimeoutCV_.notify_one();
```

Comment:
We will also want to notify if the new timeout is earlier than the current earliest.

I suppose that's not possible yet because all timeouts are probably the same... but might be worth either 
  - adding a quick comment about the current assumption that all the timeouts are the same, and to add a check there if that changes
  - or just adding the check anyway (maybe just a couple lines)

Reply:
If we wanted to implement this somewhere here, it should be something like 

```
if (futures_.size() == 1 ||
          getRPCRemainingTime(futureStartTime, futureInfo.timeout_) <
              getRPCRemainingTime(
                  futureTimeouts_.begin()->first,
                  futures_[futureTimeouts_.begin()->second.front()].timeout_)) {
                   notify()
}
```

I'm leaving just a note for now because it requires a good amount of reworking. For example, we'd also need to change the way we key into the `futureTimeouts_` map, since the key for that is the absolute start time, and we'd want to change it to something like a relative time that takes into account the specific `rpcTimeout` the RPC has.

Reply:
@jjlilley You're right for correctness, but since we have a single global RPC timeout today, they will be monotonically increasing. If we do add per-RPC timeouts (which needs to be thoroughly weighed), then we should put in additional work to avoid waking up this thread every time we add a future.

Reply:
Except if setRpcTimeout() is called after starting (e.g. cut from 1 min to 5 seconds), which is technically possible via the python bindings, though maybe not recommended/supported...

Reply:
Good point, thanks for calling this out. I'm going to make this method private for now until we can have a discussion about how/if we want to support this.

Reply:
Thanks for humoring me :)
(I've been bit by this sort of issue in the past)